README

== Instalation == 
For very very quick start add the following to LocalSettings.php:

 wfLoadExtension( 'AccessControl' )

== Config ==
First you set allow action 'view' for all users:

 $wgGroupPermissions['*']['read']            = true;

An next deny access for unauthorized users for actions:

 $wgGroupPermissions['*']['createaccount']   = false;
 $wgGroupPermissions['*']['edit']            = false;
 $wgGroupPermissions['*']['writeapi']        = false;
 $wgGroupPermissions['*']['createpage'] = false;
 $wgGroupPermissions['*']['createtalk'] = false;
 $wgGroupPermissions['*']['edit'] = false;
 $wgGroupPermissions['*']['createpage'] = false;

 // If you have installed Extension:LookupUser
 $wgGroupPermissions['*']['lookupuser'] = false;

 // If you have installed Extension:ProofreadPage
 $wgGroupPermissions['*']['pagetranslation'] = false;


CAUTION: Is strictly recommend set deny access for allmost of Special
         pages for anonymous users! I use for it function
         disableSomeSpecialPages() in my LocalSettings.php file. See:

…
 function disableSomeSpecialPages(&$list) {
     global $wgUser;

     if ( is_array( $wgUser->mRights ) && in_array( 'createpage', $wgUser->mRights ) ) {
         return true;
     } else {
         if ($wgUser->mId > 0 ) {
             return true;
         }
         unset($list['Mypage']);
         foreach( array(
             'Activeusers',
             'Allmessages',
             'Allpages',
             'Ancientpages',
             'ApiSandbox',
             'Blankpage',
             'BlockList',
             'Boilerplates',
             'Booksources',
             'BrokenRedirects',
             'CategoryTree',
             'ComparePages',
             'Contributions',
             'CreateCategory',
             'CreateClass',
             'CreateForm',
             'CreateTemplate',
             'Deadendpages',
             'DoubleRedirects',
             'ExpandTemplates',
             'Export',
             'ExportTranslations',
             'Fewestrevisions',
             'FileDuplicateSearch',
             'Forms',
             'FormEdit',
             'FormStart',
             'LanguageStats',
             'LinkSearch',
             'ListDuplicatedFiles',
             'Listfiles',
             'Listgrouprights',
             'Listredirects',
             'Listusers',
             'Log',
             'Lonelypages',
             'Longpages',
             'ManageMessageGroups',
             'MediaStatistics',
             'MessageGroupStats',
             'MIMEsearch',
             'Mostcategories',
             'Mostimages',
             'Mostinterwikis',
             'Mostlinked',
             'Mostlinkedcategories',
             'Mostlinkedtemplates',
             'Mostrevisions',
             'Movepage',
             'MultiPageEdit',
             'Newimages',
             'Newpages',
             'PagesWithProp',
             'PagesWithoutScans',
             'PageTranslation',
             'Preferences',
             'Prefixindex',
             'Protectedpages',
             'Protectedtitles',
             'RandomInCategory',
             'Randompage',
             'Randomredirect',
             'Redirect',
             'ResetTokens',
             'RunQuery',
             'SearchTranslations',
             'Shortpages',
             'Statistics',
             'SupportedLanguages',
             'TrackingCategories',
             'Translate',
             'Translations',
             'TranslationStats',
             'TranslationStash',
             'Uncategorizedcategories',
             'Uncategorizedimages',
             'Uncategorizedpages',
             'Uncategorizedtemplates',
             'Unusedimages',
             'Unusedtemplates',
             'Wantedfiles',
             'Wantedpages',
             'Wantedtemplates',
             'Watchlist',
             'Withoutinterwiki'
             ) as $i ) {
             unset( $list[$i] );
         }
     }
     return true;
 }
 $wgHooks['SpecialPage_initList'][]='disableSomeSpecialPages';
…

More documentation is at 
 https://www.mediawiki.org/wiki/Help:Extension:AccessControl

=== Global variables used by AccessControl ===
; $wgAccessControlNamespaces : Array. By default is not set. Respectively use array <code>[ 0, 12, 13 ]</code>
::0 - NS_MAIN
::12 - NS_USER
::13 - NS_USER_TALK

== How to test ==
1, Create first self-protect page and try access over link for authorized user, and next for anonymous user (is recommend use two browsers for it)

2, Try access to self-protected page by redirect.

3, Try access by transclusion self-protected page into another page

4, Try create self-protect template and use it

5, Try include page over DPL (if used it)

6, Try get content from self-protected page by find

7, Try get content from protected page by random diff

    http://server/wiki/index.php?title=blabla&diff=123

NOTE: Redirect is in beta version code deactivated!!!


changelog.txt
version 3.0.1
* Patch for deny edit access over visualeditor, proposed by user HtheChemist and tested by
  Nicolas Nallet with MediaWiki 1.34.0

version 3.0
* Completly rewrited for templates support by Aleš Kapica
* New syntax for accesslists
* Old syntax is still supported
* Used hook "ParserBeforeStrip" to secure content of pages included by DPL
* Used hook "ModifyExportQuery" to protection page before export
* Used hook "ShowSearchHit" to hide results of pages with hits
* Removed hook:
    UnknownAction (obsolete)
* New variables:
    $wgAccessControlNamespaces Array of numeric values of the namespaces with the accesslists
    $wgAccessToHistory Can allow read history or wiki content of unprotected pages for anonymous users

version 2.6
* Added support for PHP codesniffer checks, internal restructuring of code by Siebrand Mazeland
* Fixes to the above change by Thomas Mulhall

version 2.5
* Actualization of description header
* Fix for uninitialize an variable $rights in function wfFromTemplates() and repair of indents in code
* New the variable $wgAccessControlRedirect to allow set display of search results (default is false)
* New language string accesscontrol-info-box in i18n localization file

version 2.4.1
* minor fix in german translationfile
* cosmetic fixes to avoid code style errors

version 2.4.0
* Migrate to JSON i18n

verison 2.3
* Fixed function wfDoRedirect() and wfFromTemplates()
* Removed unused messages from localization file
* New function wfOnUnknownAction()

version 2.2
* Fix for MediaWiki >= 1.21 by unknown AccessControl user

version 2.1
* Fixed security problem with insert protected page, or template

version 2.0
* Completly rewrited for MediaWiki >= 1.18 by Aleš Kapica
* Removed not used variables:
    $wgAccessControlDebug
    $wgAccessControlDebugFile
    $wgAllowInfo
    $wgAllowUserList
    $wgAccessControlMessages

version 1.3
* Gabor Simon fixed problem in function wfGetUsersFromPages()
* Applying patch from Stephan Herrmann

version 1.2
* Fixed errors in parameter fuction wfDoControlUserAccess()
* Repairing while lopp in function getContentTag()

version 1.1
* Fixed security bug with accessibility on protected page over #REDIRECT

version 1.0
* Refactored all code
* Added localization file
* Changed separator on one comma (from double)
* Readonly access is can not only group, but user too

version 0.8
*Fixes for hook errors in MediaWiki 1.11.0 included
*Support for the Group of anonymous users added
*UNIQ/QUINU-Bug kind of fixed (own Link-Parser)

version 0.7
*now it's possible to have read-only groups.
*set the english language as default
*refactored some functions for better readability
*it's now possible to use the internal groups from MediaWiki
*Added a TODO-Document

version 0.6
*Bugfix release, touching the article didn't work, so I use a header redirect (bad hack, but works).
*works now also for anonymous users

version 0.5
*Articles with the accesscontrol tag are now touched in advance to displaying it, so the page will not read from the cache
*removed some development junk ;-)
*better documentation of the accesscontrolSettings.php
*added some debugging code

version 0.4
*tested in Version 1.8.2
*added some Tips
*if more then one group is in the <accesscontrol> tag, all groups are now displayed
*there is now an option, where you can turn off the behaviour, that users in the sysop group can see restricted pages
*the "access is allowed only for group..." text is now in the accesscontrolSettings.php for easier localisation

version 0.3
*access is now also controlled for editing the pages (if you access it manually per action=edit in the URL)
*only sysops can now view and edit the "Usergroup:.." pages
*added a changelog ;-)

version 0.2
*some fixes, to make it work under Version 1.7.1
*Sysops can now always access pages, so if you make an error, you have always the opportunity to correct it ;-)

version 0.1
first Version


debugHooks.php
<?php
	$wgAccesscontrolDebug = true;
	$wgAccesscontrolDebugFile = "C:/Development/testwiki/config/hooks.txt";

	$wgHooks['GetBlockedStatus'][] = 'hookGetBlockedStatus';
	$wgHooks['MagicWordMagicWords'][] = 'hookMagicWordMagicWords';
	$wgHooks['MagicWordwgVariableIDs'][] = 'hookMagicWordwgVariableIDs';
	$wgHooks['OutputPageParserOutput'][] = 'hookOutputPageParserOutput';
	$wgHooks['ParserAfterStrip'][] = 'hookParserAfterStrip';
	$wgHooks['ParserAfterTidy'][] = 'hookParserAfterTidy';
	$wgHooks['ParserBeforeInternalParse'][] = 'hookParserBeforeInternalParse';
	$wgHooks['ParserBeforeTidy'][] = 'hookParserBeforeTidy';
	$wgHooks['ParserClearState'][] = 'hookParserClearState';
	$wgHooks['ParserGetVariableValueSwitch'][] = 'hookParserGetVariableValueSwitch';
	$wgHooks['ParserGetVariableValueTs'][] = 'hookParserGetVariableValueTs';
	$wgHooks['ParserGetVariableValueVarCache'][] = 'hookParserGetVariableValueVarCache';
	$wgHooks['userCan'][] = 'hookuserCan2';


	function debugme($input)
	{
		global $wgAccesscontrolDebug;
		global $wgAccesscontrolDebugFile;

		if ($wgAccesscontrolDebug)
		{
			$f = fopen($wgAccesscontrolDebugFile, "a+");
			fputs($f, $input."\r\n");
			fclose($f);
		}
	}

	function hookGetBlockedStatus() { debugme( "hookGetBlockedStatus" ); }
	function hookMagicWordMagicWords() { debugme( "hookMagicWordMagicWords" ); }
	function hookMagicWordwgVariableIDs() { debugme( "hookMagicWordwgVariableIDs" ); }
	function hookOutputPageParserOutput(&$outputPage, $parserOutput) 
	{ 
		debugme( "hookOutputPageParserOutput" ); 
		debugme( $parserOutput->mText );
	}
	function hookParserAfterStrip() { debugme( "hookParserAfterStrip" ); }
	function hookParserAfterTidy() { debugme( "hookParserAfterTidy" ); }
	function hookParserBeforeInternalParse() { debugme( "hookParserBeforeInternalParse" ); }
	function hookParserBeforeTidy() { debugme( "hookParserBeforeTidy" ); }
	function hookParserClearState() { debugme( "hookParserClearState" ); }
	function hookParserGetVariableValueSwitch() { debugme( "hookParserGetVariableValueSwitch" ); }
	function hookParserGetVariableValueTs() { debugme( "hookParserGetVariableValueTs" ); }
	function hookParserGetVariableValueVarCache() { debugme( "hookParserGetVariableValueVarCache" ); }
	function hookRecentChange_save() { debugme( "hookRecentChange_save" ); }
	function hookuserCan2() { debugme( "hookuserCan" ); }
?>


AccessControl.i18n.php
<?php
/**
 * Internationalisation file for AccessControl extension.
 *
 * @addtogroup Extensions
 */

$messages = array();

$messages['en'] = array(
	'accesscontrol-desc' => 'This is extension that enables group access restriction on a page by user',
	'accesscontrol-group' => 'This page is only accessible for group $1 !!!',
	'accesscontrol-groups' => 'This page is only accessible for the groups $1 !!!',
	'accesscontrol-info-user' => 'Only_sysop',
	'accesscontrol-info-anonymous' => 'No_anonymous',
	'accesscontrol-info-deny' => 'No_Access',
	'accesscontrol-edit-anonymous' => 'Deny_anonymous',
	'accesscontrol-edit-users' => 'Deny_edit_list',
);
/** Czech (Česky)
 * @author Aleš Kapica
 */
$messages['cs'] = array(
	'accesscontrol-desc' => 'Toto je rozšíření, které přidává uživatelskou možnost omezení přístupu ke stránce',
	'accesscontrol-group' => 'Tato stránka je přístupná pouze pro skupinu $1 !!!',
	'accesscontrol-groups' => 'Tato stránka je přístupná pouze pro skupiny $1 !!!',
	'accesscontrol-info-user' => 'Stránka má omezený přístup k administraci',
	'accesscontrol-info-anonymous' => 'Ke stránce je zakázán anonymní přístup',
	'accesscontrol-info-deny' => 'Stránka má omezený přístup',
	'accesscontrol-edit-anonymous' => 'Jen pro registrované',
	'accesscontrol-edit-users' => 'Stránka jen pro správce',
);


AccessControl.php
<?php

/* MediaWiki extension that enables group access restriction on a page-by-page
 * basis contributed by Martin Mueller (http://blog.pagansoft.de) based into 
 * version 1.3 on accesscontrol.php by Josh Greenberg.
 * Version 2.0 for MediaWiki >= 1.18 rewrited completly by Aleš Kapica.
 * @package MediaWiki
 * @subpackage Extensions
 * @author Aleš Kapica
 * @copyright 2008-2012 Aleš Kapica
 * @licence GNU General Public Licence
 */

if( !defined( 'MEDIAWIKI' ) ) {
	echo ( "This file is an extension to the MediaWiki software and cannot be used standalone.\n" );
	die();
}

$wgAdminCanReadAll = true;				// sysop users can read all restricted pages

$wgExtensionCredits['specialpage']['AccessControl'] = array(
	'name'			=> 'AccessControlExtension',
	'author'		=> array( 'Aleš Kapica' ),
	'url'			=> 'http://www.mediawiki.org/wiki/Extension:AccessControl',
	'version'		=> '2.0',
	'description'		=> 'Access control based on users lists. Administrator rights need not be for it.',
	'descriptionmsg'	=> 'accesscontrol-desc',
);

$wgHooks['ParserFirstCallInit'][] = 'wfAccessControlExtension' ;

$dir = dirname(__FILE__) . '/';
$wgExtensionMessagesFiles['AccessControl'] = $dir . 'AccessControl.i18n.php';


//Hook the userCan function for bypassing the cache
$wgHooks['userCan'][] = 'hookUserCan';

function wfAccessControlExtension( Parser $parser ) {
	/* This the hook function adds the tag <accesscontrol> to the wiki parser */
	$parser->setHook( "accesscontrol", "doControlUserAccess" );
	return true;
}

function doControlUserAccess( $input, array $args, Parser $parser, PPFrame $frame) {
	/* Funcion called by wfAccessControlExtension */
	return displayGroups();
}

function accessControl($obsahtagu){
	$accessgroup = Array( Array(), Array() );
	$listaccesslist = explode(",",$obsahtagu);
	foreach ( $listaccesslist as $accesslist ) {
		if (strpos($accesslist,"(ro)") !== false) {
			$accesslist = trim(str_replace("(ro)","",$accesslist));
			$group = makeGroupArray($accesslist);
			$accessgroup[1] = array_merge($accessgroup[1] , $group[0]);
			$accessgroup[1] = array_merge($accessgroup[1] , $group[1]);
			} else {
			$accesslist = trim($accesslist);
			$group = makeGroupArray($accesslist);
			$accessgroup[0] = array_merge($accessgroup[0] , $group[0]);
			$accessgroup[1] = array_merge($accessgroup[1] , $group[1]);
			}
		}
	return $accessgroup;
}

function makeGroupArray($accesslist) {
	/* Function returns array with two lists.
		First is list full access users.
		Second is list readonly users. */
	$userswrite = Array();
	$usersreadonly = Array();
		$users=getUsersFromPages($accesslist);
		foreach(array_keys($users) as $user ) {
			switch ($users[$user]) {
				case 'read':
					$usersreadonly[] = $user;
					break;
				case 'edit':
					$userswrite[] = $user;
					break;
				}
			}
	return array( $userswrite , $usersreadonly );
}

function displayGroups() {
	/* Function replace the tag <accesscontrol> and his content, behind info about a protection this the page */
	$style = "<p id=\"accesscontrol\" style=\"text-align:center;color:#BA0000;font-size:8pt\">";
	$text = wfMsg('accesscontrol-info');
	$style_end = "</p>";
	$wgAllowInfo = $style.$text.$style_end;
	return $wgAllowInfo;
}

function getContentPage(&$title) {
	/* Function get content the page identified by title object
       from database */
	$Title = new Title();
	$gt = $Title->makeTitle(0, $title->mDbkeyform);
	// create Article and get the content
	$contentPage = new Article( $gt, 0 );
	return $contentPage->fetchContent(0);
	}

function getUsersFromPages($skupina) {
	/* Extracts the allowed users from the userspace access list */
	$allowedAccess = Array();
	$allow = Array();
	$Title = new Title();
	$gt = $Title->makeTitle(0, $skupina);
	// create Article and get the content
	$groupPage = new Article( $gt, 0 );
	$allowedUsers=$groupPage->fetchContent(0);
	$groupPage = NULL;
	$usersAccess = explode("\n", $allowedUsers);
	foreach ($usersAccess as $userEntry) {
		$userItem = trim($userEntry);
		if (substr($userItem,0,1) == "*") {
			if (strpos($userItem,"(ro)") === false) {
				$user = trim(str_replace("*","",$userItem));
				$allow[$user] = 'edit';
			} else {
				$user = trim(str_replace("*","",$userItem));
				$user = trim(str_replace("(ro)","",$user));
				$allow[$user] = 'read';
			}
		}
	}
	if (is_array($allow)) {
		$allowedAccess = $allow;
		unset($allow);
	}
	return $allowedAccess;
}

function doRedirect($info) {
	/* make redirection for non authorized users */
	global $wgScript, $wgSitename, $wgOut;

	if (!$info) $info="No_access";
	if ($info == "Only_sysop") {
		$target = wfMsg('accesscontrol-info-user');
	} elseif ($info == "No_anonymous") {
		$target = wfMsg('accesscontrol-info-anonymous');
	} elseif ($info == "Deny_anonymous") {
		$target = wfMsg('accesscontrol-edit-anonymous');
	} elseif ($info == "Deny_edit_list") {
		$target = wfMsg('accesscontrol-edit-users');
	} else {
		$target = wfMsg('accesscontrol-info-deny');
	}
	if (isset($_SESSION['redirect'])) {
		// removing info about redirect from session after move..
		unset($_SESSION['redirect']);
	}
	header("Location: ".$wgScript."/".$wgSitename.":".$target);
}

function allRightTags($string) {
	/* Function for extraction content tag accesscontrol from raw source the page */
	$contenttag=Array();
	$starttag = "<accesscontrol>";
	$endtag = "</accesscontrol>";
	$redirecttag = "redirect";

	if ((mb_substr(trim($string), 0, 1) == "#") && (stripos(mb_substr(trim($string), 1, 9), $redirecttag) == "0")) {
		/* Treatment redirects - content variable $string must be replaced over content the target page */
		$sourceredirecttag=mb_substr($string, 0, strpos($string,']]'));
		$redirecttarget=trim(substr($sourceredirecttag, strpos($sourceredirecttag,'[[')+2));
		if (strpos($redirecttarget,'|')) {
			$redirecttarget=trim(substr($redirecttarget, 0, strpos($redirecttarget,'|')));
			}
		$Title = new Title();
		$gt = $Title->makeTitle(0, $redirecttarget);
		return allRightTags(getContentPage($gt));
		}

	$start = strpos( $string, $starttag );
	if ($start !== false) {
		$start += strlen($starttag);
		$end = strpos( $string, $endtag );
		if ($end !== false) {
			$groupsString = substr($string, $start, $end-$start );
			if (strlen($groupsString) == 0) {
				$contenttag['end'] = strlen($starttag)+strlen($endtag); 
			} else {
				$contenttag['groups']=$groupsString;
				$contenttag['end'] = $end+strlen($endtag);
			}

			if(isset($_SESSION['redirect'])) {
				$_SESSION['redirect'] = $contenttag;
			} else {
				return $contenttag;
			}
		}
	} else {
		if(isset($_SESSION['redirect'])) {
			return $_SESSION['redirect'];
		} else {
			return false;
			}
	}
}

function hookUserCan(&$title, &$wgUser, $action, &$result) {
	/* Main function control access for all users */
	global $wgActions, $wgAdminCanReadAll;
	if ( $wgUser->mId === 0 ) {
		/* Deny actions for all anonymous */
		$wgActions['edit']=false;
		$wgActions['history']=false;
		$wgActions['submit']=false;
		$wgActions['info']=false;
		$wgActions['raw']=false;
		$wgActions['delete']=false;
		$wgActions['revert']=false;
		$wgActions['revisiondelete']=false;
		$wgActions['rollback']=false;
		$wgActions['markpatrolled']=false;
		}
	$rights=allRightTags(getContentPage($title));
	if (is_array($rights)) {
		if ( $wgUser->mId === 0 ) {
			/* Redirection unknown users */
			$wgActions['view']=false;
			doRedirect('accesscontrol-info-anonymous');
			} else {
				if (in_array('sysop',$wgUser->mGroups, true)) {
					if (isset($wgAdminCanReadAll)) {
						if ($wgAdminCanReadAll) {
							return true;
						}
					}
				}				
				$users=accessControl($rights['groups']);
				if (in_array($wgUser->mName, $users[0], true)) {
						return true;
					} else {
						$wgActions['edit']=false;
						$wgActions['history']=false;
						$wgActions['submit']=false;
						$wgActions['info']=false;
						$wgActions['raw']=false;
						$wgActions['delete']=false;
						$wgActions['revert']=false;
						$wgActions['revisiondelete']=false;
						$wgActions['rollback']=false;
						$wgActions['markpatrolled']=false;
						if (in_array($wgUser->mName, $users[1], true)) {
								return true;
							} else {
								$wgActions['view']=false;
								return doRedirect('accesscontrol-info-anonymous');
							}
					}
			}
		} else {
			return true;
		}
}

?>
